File: java/219.contains-duplicate-ii.java
/*
 * @lc app=leetcode id=219 lang=java
 *
 * [219] Contains Duplicate II
 */
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        if (nums == null) return false;
        Set<Integer> lookup = new HashSet<>();
        for (int r = 0; r < nums.length; ++r) {
            if (r-k-1 >= 0) lookup.remove(nums[r-k-1]);
            if (lookup.contains(nums[r])) return true;
            lookup.add(nums[r]);
        }
        return false;
    }
}

File: go/219.contains-duplicate-ii.go
/*
 * @lc app=leetcode id=219 lang=golang
 *
 * [219] Contains Duplicate II
 */

// @lc code=start

func containsNearbyDuplicate(nums []int, k int) bool {
	cache := make(map[int]struct{})
	for i := 0; i < len(nums); i++ {
		if _, b := cache[nums[i]]; b {
			return true
		}
		cache[nums[i]] = struct{}{}
		if i >= k {
			delete(cache, nums[i-k])
		}
	}
	return false
}

// @lc code=end

